Add more constants, functions, and tests for `f16` and `f128`

tgross35 opened the pull request:
This adds everything that was in some way blocked on const eval, since https://github.com/rust-lang/rust/pull/126429 landed. There is a lot of `cfg(bootstrap)` since that is a fairly recent change.

`f128` tests are disabled on everything except x86_64 and Linux aarch64, which are two platforms I know have "good" support for these types - meaning basic math symbols are available and LLVM doesn't hit selection crashes. `f16` tests are enabled on almost everything except for known LLVM crashes. Doctests are only enabled on x86_64.

Tracking issue: https://github.com/rust-lang/rust/issues/116909

tgross35 committed:
Extract repeated constants from `f32` and `f64` source

This will make it easier to keep `f16` and `f128` consistent as their
implementations get added.

tgross35 commented:
@rustbot author

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-17`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9554143339/job/26334569452) [(plain)](https://github.com/rust-lang/rust/commit/76797067048de3cf65fae7010a9fa31762a70930/checks/26334569452/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#16 exporting to docker image format
#16 sending tarball 25.0s done
#16 DONE 30.4s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-17]
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-17', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-17/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: change-id            := 99999999
---
  Downloaded boml v0.3.1
   Compiling boml v0.3.1
   Compiling y v0.1.0 (/checkout/compiler/rustc_codegen_gcc/build_system)
    Finished `release` profile [optimized] target(s) in 3.59s
     Running `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-codegen/x86_64-unknown-linux-gnu/release/y test --use-system-gcc --use-backend gcc --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc --release --mini-tests --std-tests`
Using system GCC
[BUILD] example
[AOT] mini_core_hello_world
/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc/mini_core_hello_world
abc
---
Test executable failed (exit status: 101).

stderr:
thread 'main' panicked at library/core/src/num/f128.rs:28:5:
assertion failed: bois.into_iter().map(|b|
                b.weight).zip([-5.0, 0.1, 10.0, 99.0, f128::INFINITY,
                    f128::NAN].iter()).all(|(a, b)| a.to_bits() == b.to_bits())



failures:
failures:
    library/core/src/num/f128.rs - f128::f128::next_up (line 394)
    library/core/src/num/f128.rs - f128::f128::total_cmp (line 786)

test result: FAILED. 4656 passed; 2 failed; 43 ignored; 0 measured; 0 filtered out; finished in 56.48s

error: doctest failed, to rerun pass `-p core --doc`
  local time: Mon Jun 17 20:41:00 UTC 2024
  network time: Mon, 17 Jun 2024 20:41:00 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9557715530/job/26345251551) [(plain)](https://github.com/rust-lang/rust/commit/1df61e16fb60fcd8e01e848d402391eb93e50f8b/checks/26345251551/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
                               at /rustc/75ac3b6331873133c4f7a10f2252afd6f3906c6a/library/std/src/sys_common/backtrace.rs:34:9
   9:     0x7f6bd098d729 - std::panicking::default_hook::{{closure}}::hc8d23e37f5c72049
  10:     0x7f6bd098d4ca - std::panicking::default_hook::he5fe33ec98ad6aeb
                               at /rustc/75ac3b6331873133c4f7a10f2252afd6f3906c6a/library/std/src/panicking.rs:298:9
  11:     0x7f6bd3a4bb50 - std[1d26776ff6539ee1]::panicking::update_hook::<alloc[a18d69858d21b7b]::boxed::Box<rustc_driver_impl[f9bf230c2e084b5c]::install_ice_hook::{closure#0}>>::{closure#0}
  12:     0x7f6bd098de5b - <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call::h2ffbb0ba3b07d310
  13:     0x7f6bd098de5b - std::panicking::rust_panic_with_hook::hfe10a8a9e06dddff
                               at /rustc/75ac3b6331873133c4f7a10f2252afd6f3906c6a/library/std/src/panicking.rs:799:13
  14:     0x7f6bd098db9b - std::panicking::begin_panic_handler::{{closure}}::h99216258a68d8f5a
                               at /rustc/75ac3b6331873133c4f7a10f2252afd6f3906c6a/library/std/src/panicking.rs:656:13
                               at /rustc/75ac3b6331873133c4f7a10f2252afd6f3906c6a/library/std/src/panicking.rs:656:13
  15:     0x7f6bd098b3d9 - std::sys_common::backtrace::__rust_end_short_backtrace::h49a766e96193ca7e
                               at /rustc/75ac3b6331873133c4f7a10f2252afd6f3906c6a/library/std/src/sys_common/backtrace.rs:171:18
  16:     0x7f6bd098d907 - rust_begin_unwind
                               at /rustc/75ac3b6331873133c4f7a10f2252afd6f3906c6a/library/std/src/panicking.rs:652:5
  17:     0x7f6bd09d81f3 - core::panicking::panic_fmt::hcb6f6b919b498d25
                               at /rustc/75ac3b6331873133c4f7a10f2252afd6f3906c6a/library/core/src/panicking.rs:72:14
  18:     0x5593a2a2b047 - clippy_utils[8a4c222704ae5b46]::consts::lit_to_mir_constant
  19:     0x5593a2a2c394 - <clippy_utils[8a4c222704ae5b46]::consts::ConstEvalLateContext>::expr
  20:     0x5593a2a2b230 - clippy_utils[8a4c222704ae5b46]::consts::constant_simple
  21:     0x5593a2924627 - <clippy_lints[87b25297f1b3f31b]::zero_div_zero::ZeroDiv as rustc_lint[101d3b3dbf684c39]::passes::LateLintPass>::check_expr
  22:     0x7f6bd3e1dcce - <rustc_lint[101d3b3dbf684c39]::late::LateContextAndPass<rustc_lint[101d3b3dbf684c39]::late::RuntimeCombinedLateLintPass> as rustc_hir[90aed0b98244eff2]::intravisit::Visitor>::visit_expr::{closure#0}::{closure#0}
  23:     0x7f6bd3e1dc31 - <rustc_lint[101d3b3dbf684c39]::late::LateContextAndPass<rustc_lint[101d3b3dbf684c39]::late::RuntimeCombinedLateLintPass> as rustc_hir[90aed0b98244eff2]::intravisit::Visitor>::visit_expr
  24:     0x7f6bd3e1e648 - <rustc_lint[101d3b3dbf684c39]::late::LateContextAndPass<rustc_lint[101d3b3dbf684c39]::late::RuntimeCombinedLateLintPass> as rustc_hir[90aed0b98244eff2]::intravisit::Visitor>::visit_block
  25:     0x7f6bd3e1e4a7 - <rustc_lint[101d3b3dbf684c39]::late::LateContextAndPass<rustc_lint[101d3b3dbf684c39]::late::RuntimeCombinedLateLintPass> as rustc_hir[90aed0b98244eff2]::intravisit::Visitor>::visit_expr::{closure#0}::{closure#0}
  26:     0x7f6bd3e1dc31 - <rustc_lint[101d3b3dbf684c39]::late::LateContextAndPass<rustc_lint[101d3b3dbf684c39]::late::RuntimeCombinedLateLintPass> as rustc_hir[90aed0b98244eff2]::intravisit::Visitor>::visit_expr
  27:     0x7f6bd3e1da38 - <rustc_lint[101d3b3dbf684c39]::late::LateContextAndPass<rustc_lint[101d3b3dbf684c39]::late::RuntimeCombinedLateLintPass> as rustc_hir[90aed0b98244eff2]::intravisit::Visitor>::visit_nested_body
  28:     0x7f6bd3e1e5d4 - <rustc_lint[101d3b3dbf684c39]::late::LateContextAndPass<rustc_lint[101d3b3dbf684c39]::late::RuntimeCombinedLateLintPass> as rustc_hir[90aed0b98244eff2]::intravisit::Visitor>::visit_fn
  29:     0x7f6bd3e09ca5 - <rustc_lint[101d3b3dbf684c39]::late::LateContextAndPass<rustc_lint[101d3b3dbf684c39]::late::RuntimeCombinedLateLintPass> as rustc_hir[90aed0b98244eff2]::intravisit::Visitor>::visit_nested_item
  30:     0x7f6bd3e08d48 - <rustc_lint[101d3b3dbf684c39]::late::LateContextAndPass<rustc_lint[101d3b3dbf684c39]::late::RuntimeCombinedLateLintPass> as rustc_hir[90aed0b98244eff2]::intravisit::Visitor>::visit_nested_item
  31:     0x7f6bd5808bcd - rustc_lint[101d3b3dbf684c39]::late::check_crate
  32:     0x7f6bd580cfb4 - rustc_interface[df29f1c958d14d7d]::passes::analysis
  33:     0x7f6bd580c125 - rustc_query_impl[28192108fdaf43fb]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[28192108fdaf43fb]::query_impl::analysis::dynamic_query::{closure#2}::{closure#0}, rustc_middle[953e3de138f4a527]::query::erase::Erased<[u8; 1usize]>>
  34:     0x7f6bd5b58da2 - rustc_query_system[560c854d2bb8284e]::query::plumbing::try_execute_query::<rustc_query_impl[28192108fdaf43fb]::DynamicConfig<rustc_query_system[560c854d2bb8284e]::query::caches::SingleCache<rustc_middle[953e3de138f4a527]::query::erase::Erased<[u8; 1usize]>>, false, false, false>, rustc_query_impl[28192108fdaf43fb]::plumbing::QueryCtxt, false>
  35:     0x7f6bd5b58b4f - rustc_query_impl[28192108fdaf43fb]::query_impl::analysis::get_query_non_incr::__rust_end_short_backtrace
  36:     0x7f6bd5a50c52 - rustc_interface[df29f1c958d14d7d]::interface::run_compiler::<core[ec4a210f237120ac]::result::Result<(), rustc_span[c72166897846c570]::ErrorGuaranteed>, rustc_driver_impl[f9bf230c2e084b5c]::run_compiler::{closure#0}>::{closure#1}
  37:     0x7f6bd58d8e0b - std[1d26776ff6539ee1]::sys_common::backtrace::__rust_begin_short_backtrace::<rustc_interface[df29f1c958d14d7d]::util::run_in_thread_with_globals<rustc_interface[df29f1c958d14d7d]::interface::run_compiler<core[ec4a210f237120ac]::result::Result<(), rustc_span[c72166897846c570]::ErrorGuaranteed>, rustc_driver_impl[f9bf230c2e084b5c]::run_compiler::{closure#0}>::{closure#1}, core[ec4a210f237120ac]::result::Result<(), rustc_span[c72166897846c570]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[ec4a210f237120ac]::result::Result<(), rustc_span[c72166897846c570]::ErrorGuaranteed>>
  38:     0x7f6bd58d8c06 - <<std[1d26776ff6539ee1]::thread::Builder>::spawn_unchecked_<rustc_interface[df29f1c958d14d7d]::util::run_in_thread_with_globals<rustc_interface[df29f1c958d14d7d]::interface::run_compiler<core[ec4a210f237120ac]::result::Result<(), rustc_span[c72166897846c570]::ErrorGuaranteed>, rustc_driver_impl[f9bf230c2e084b5c]::run_compiler::{closure#0}>::{closure#1}, core[ec4a210f237120ac]::result::Result<(), rustc_span[c72166897846c570]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[ec4a210f237120ac]::result::Result<(), rustc_span[c72166897846c570]::ErrorGuaranteed>>::{closure#2} as core[ec4a210f237120ac]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  39:     0x7f6bd0997e4b - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::he55b257296120b17
                               at /rustc/75ac3b6331873133c4f7a10f2252afd6f3906c6a/library/alloc/src/boxed.rs:2063:9
  40:     0x7f6bd0997e4b - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::hab42474581aa3d2c
  41:     0x7f6bd0997e4b - std::sys::pal::unix::thread::Thread::new::thread_start::h6343e8231a58f36e
                               at /rustc/75ac3b6331873133c4f7a10f2252afd6f3906c6a/library/std/src/sys/pal/unix/thread.rs:108:17
  42:     0x7f6bd0658ac3 - <unknown>
  43:     0x7f6bd06ea850 - <unknown>
  43:     0x7f6bd06ea850 - <unknown>
  44:                0x0 - <unknown>

error: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust-clippy/issues/new?template=ice.yml
note: please make sure that you have updated to the latest nightly


warning: the ICE couldn't be written to `/checkout/rustc-ice-2024-06-18T02_11_28-19295.txt`: Read-only file system (os error 30)
note: rustc 1.80.0-beta.1 (75ac3b633 2024-06-10) running on x86_64-unknown-linux-gnu


note: compiler flags: --crate-type lib -C opt-level=3 -C embed-bitcode=no -C codegen-units=1 -C debug-assertions=on -C symbol-mangling-version=legacy -Z unstable-options -Z unstable-options -Z force-unstable-if-unmarked -Z macro-backtrace -C split-debuginfo=off -C prefer-dynamic -Z validate-mir -Z mir-opt-level=3 -Z inline-mir -Z inline-mir-preserve-debug -C link-args=-Wl,-z,origin -C link-args=-Wl,-rpath,$ORIGIN/../lib -C force-frame-pointers=yes -Z crate-attr=doc(html_root_url="https://doc.rust-lang.org/nightly/") -Z binary-dep-depinfo -Z force-unstable-if-unmarked
note: some of the compiler flags provided by cargo are hidden

query stack during panic:
#0 [analysis] running analysis passes on this crate
```

</details>

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9558710957/job/26347843323) [(plain)](https://github.com/rust-lang/rust/commit/aa684b5d9f9d4d8547a57b72484ca8628f1f9fd0/checks/26347843323/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
                               at /rustc/75ac3b6331873133c4f7a10f2252afd6f3906c6a/library/std/src/sys_common/backtrace.rs:34:9
   9:     0x7f3486f8d729 - std::panicking::default_hook::{{closure}}::hc8d23e37f5c72049
  10:     0x7f3486f8d4ca - std::panicking::default_hook::he5fe33ec98ad6aeb
                               at /rustc/75ac3b6331873133c4f7a10f2252afd6f3906c6a/library/std/src/panicking.rs:298:9
  11:     0x7f348a04bb50 - std[1d26776ff6539ee1]::panicking::update_hook::<alloc[a18d69858d21b7b]::boxed::Box<rustc_driver_impl[f9bf230c2e084b5c]::install_ice_hook::{closure#0}>>::{closure#0}
  12:     0x7f3486f8de5b - <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call::h2ffbb0ba3b07d310
  13:     0x7f3486f8de5b - std::panicking::rust_panic_with_hook::hfe10a8a9e06dddff
                               at /rustc/75ac3b6331873133c4f7a10f2252afd6f3906c6a/library/std/src/panicking.rs:799:13
  14:     0x7f3486f8db9b - std::panicking::begin_panic_handler::{{closure}}::h99216258a68d8f5a
                               at /rustc/75ac3b6331873133c4f7a10f2252afd6f3906c6a/library/std/src/panicking.rs:656:13
                               at /rustc/75ac3b6331873133c4f7a10f2252afd6f3906c6a/library/std/src/panicking.rs:656:13
  15:     0x7f3486f8b3d9 - std::sys_common::backtrace::__rust_end_short_backtrace::h49a766e96193ca7e
                               at /rustc/75ac3b6331873133c4f7a10f2252afd6f3906c6a/library/std/src/sys_common/backtrace.rs:171:18
  16:     0x7f3486f8d907 - rust_begin_unwind
                               at /rustc/75ac3b6331873133c4f7a10f2252afd6f3906c6a/library/std/src/panicking.rs:652:5
  17:     0x7f3486fd81f3 - core::panicking::panic_fmt::hcb6f6b919b498d25
                               at /rustc/75ac3b6331873133c4f7a10f2252afd6f3906c6a/library/core/src/panicking.rs:72:14
  18:     0x565273528047 - clippy_utils[8a4c222704ae5b46]::consts::lit_to_mir_constant
  19:     0x565273529394 - <clippy_utils[8a4c222704ae5b46]::consts::ConstEvalLateContext>::expr
  20:     0x565273528230 - clippy_utils[8a4c222704ae5b46]::consts::constant_simple
  21:     0x565273421627 - <clippy_lints[87b25297f1b3f31b]::zero_div_zero::ZeroDiv as rustc_lint[101d3b3dbf684c39]::passes::LateLintPass>::check_expr
  22:     0x7f348a41dcce - <rustc_lint[101d3b3dbf684c39]::late::LateContextAndPass<rustc_lint[101d3b3dbf684c39]::late::RuntimeCombinedLateLintPass> as rustc_hir[90aed0b98244eff2]::intravisit::Visitor>::visit_expr::{closure#0}::{closure#0}
  23:     0x7f348a41dc31 - <rustc_lint[101d3b3dbf684c39]::late::LateContextAndPass<rustc_lint[101d3b3dbf684c39]::late::RuntimeCombinedLateLintPass> as rustc_hir[90aed0b98244eff2]::intravisit::Visitor>::visit_expr
  24:     0x7f348a41e648 - <rustc_lint[101d3b3dbf684c39]::late::LateContextAndPass<rustc_lint[101d3b3dbf684c39]::late::RuntimeCombinedLateLintPass> as rustc_hir[90aed0b98244eff2]::intravisit::Visitor>::visit_block
  25:     0x7f348a41e4a7 - <rustc_lint[101d3b3dbf684c39]::late::LateContextAndPass<rustc_lint[101d3b3dbf684c39]::late::RuntimeCombinedLateLintPass> as rustc_hir[90aed0b98244eff2]::intravisit::Visitor>::visit_expr::{closure#0}::{closure#0}
  26:     0x7f348a41dc31 - <rustc_lint[101d3b3dbf684c39]::late::LateContextAndPass<rustc_lint[101d3b3dbf684c39]::late::RuntimeCombinedLateLintPass> as rustc_hir[90aed0b98244eff2]::intravisit::Visitor>::visit_expr
  27:     0x7f348a41da38 - <rustc_lint[101d3b3dbf684c39]::late::LateContextAndPass<rustc_lint[101d3b3dbf684c39]::late::RuntimeCombinedLateLintPass> as rustc_hir[90aed0b98244eff2]::intravisit::Visitor>::visit_nested_body
  28:     0x7f348a41e5d4 - <rustc_lint[101d3b3dbf684c39]::late::LateContextAndPass<rustc_lint[101d3b3dbf684c39]::late::RuntimeCombinedLateLintPass> as rustc_hir[90aed0b98244eff2]::intravisit::Visitor>::visit_fn
  29:     0x7f348a409ca5 - <rustc_lint[101d3b3dbf684c39]::late::LateContextAndPass<rustc_lint[101d3b3dbf684c39]::late::RuntimeCombinedLateLintPass> as rustc_hir[90aed0b98244eff2]::intravisit::Visitor>::visit_nested_item
  30:     0x7f348a408d48 - <rustc_lint[101d3b3dbf684c39]::late::LateContextAndPass<rustc_lint[101d3b3dbf684c39]::late::RuntimeCombinedLateLintPass> as rustc_hir[90aed0b98244eff2]::intravisit::Visitor>::visit_nested_item
  31:     0x7f348be08bcd - rustc_lint[101d3b3dbf684c39]::late::check_crate
  32:     0x7f348be0cfb4 - rustc_interface[df29f1c958d14d7d]::passes::analysis
  33:     0x7f348be0c125 - rustc_query_impl[28192108fdaf43fb]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[28192108fdaf43fb]::query_impl::analysis::dynamic_query::{closure#2}::{closure#0}, rustc_middle[953e3de138f4a527]::query::erase::Erased<[u8; 1usize]>>
  34:     0x7f348c158da2 - rustc_query_system[560c854d2bb8284e]::query::plumbing::try_execute_query::<rustc_query_impl[28192108fdaf43fb]::DynamicConfig<rustc_query_system[560c854d2bb8284e]::query::caches::SingleCache<rustc_middle[953e3de138f4a527]::query::erase::Erased<[u8; 1usize]>>, false, false, false>, rustc_query_impl[28192108fdaf43fb]::plumbing::QueryCtxt, false>
  35:     0x7f348c158b4f - rustc_query_impl[28192108fdaf43fb]::query_impl::analysis::get_query_non_incr::__rust_end_short_backtrace
  36:     0x7f348c050c52 - rustc_interface[df29f1c958d14d7d]::interface::run_compiler::<core[ec4a210f237120ac]::result::Result<(), rustc_span[c72166897846c570]::ErrorGuaranteed>, rustc_driver_impl[f9bf230c2e084b5c]::run_compiler::{closure#0}>::{closure#1}
  37:     0x7f348bed8e0b - std[1d26776ff6539ee1]::sys_common::backtrace::__rust_begin_short_backtrace::<rustc_interface[df29f1c958d14d7d]::util::run_in_thread_with_globals<rustc_interface[df29f1c958d14d7d]::interface::run_compiler<core[ec4a210f237120ac]::result::Result<(), rustc_span[c72166897846c570]::ErrorGuaranteed>, rustc_driver_impl[f9bf230c2e084b5c]::run_compiler::{closure#0}>::{closure#1}, core[ec4a210f237120ac]::result::Result<(), rustc_span[c72166897846c570]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[ec4a210f237120ac]::result::Result<(), rustc_span[c72166897846c570]::ErrorGuaranteed>>
  38:     0x7f348bed8c06 - <<std[1d26776ff6539ee1]::thread::Builder>::spawn_unchecked_<rustc_interface[df29f1c958d14d7d]::util::run_in_thread_with_globals<rustc_interface[df29f1c958d14d7d]::interface::run_compiler<core[ec4a210f237120ac]::result::Result<(), rustc_span[c72166897846c570]::ErrorGuaranteed>, rustc_driver_impl[f9bf230c2e084b5c]::run_compiler::{closure#0}>::{closure#1}, core[ec4a210f237120ac]::result::Result<(), rustc_span[c72166897846c570]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[ec4a210f237120ac]::result::Result<(), rustc_span[c72166897846c570]::ErrorGuaranteed>>::{closure#2} as core[ec4a210f237120ac]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  39:     0x7f3486f97e4b - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::he55b257296120b17
                               at /rustc/75ac3b6331873133c4f7a10f2252afd6f3906c6a/library/alloc/src/boxed.rs:2063:9
  40:     0x7f3486f97e4b - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::hab42474581aa3d2c
  41:     0x7f3486f97e4b - std::sys::pal::unix::thread::Thread::new::thread_start::h6343e8231a58f36e
                               at /rustc/75ac3b6331873133c4f7a10f2252afd6f3906c6a/library/std/src/sys/pal/unix/thread.rs:108:17
  42:     0x7f3486d3fac3 - <unknown>
  43:     0x7f3486dd1850 - <unknown>
  43:     0x7f3486dd1850 - <unknown>
  44:                0x0 - <unknown>

error: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust-clippy/issues/new?template=ice.yml
note: please make sure that you have updated to the latest nightly


warning: the ICE couldn't be written to `/checkout/rustc-ice-2024-06-18T04_06_08-19291.txt`: Read-only file system (os error 30)
note: rustc 1.80.0-beta.1 (75ac3b633 2024-06-10) running on x86_64-unknown-linux-gnu


note: compiler flags: --crate-type lib -C opt-level=3 -C embed-bitcode=no -C codegen-units=1 -C debug-assertions=on -C symbol-mangling-version=legacy -Z unstable-options -Z unstable-options -Z force-unstable-if-unmarked -Z macro-backtrace -C split-debuginfo=off -C prefer-dynamic -Z validate-mir -Z mir-opt-level=3 -Z inline-mir -Z inline-mir-preserve-debug -C link-args=-Wl,-z,origin -C link-args=-Wl,-rpath,$ORIGIN/../lib -C force-frame-pointers=yes -Z crate-attr=doc(html_root_url="https://doc.rust-lang.org/nightly/") -Z binary-dep-depinfo -Z force-unstable-if-unmarked
note: some of the compiler flags provided by cargo are hidden

query stack during panic:
#0 [analysis] running analysis passes on this crate
```

</details>

tgross35 commented:
@rustbot label +rla-silenced

tgross35 commented:
Ah... guess the Clippy ICEs need to be fixed before this can be merged, `#[allow(clippy)]` still hits the ICE. Bit of an unfortunate sequence but I guess I need to fix that anyway.

tspiteri commented on library/core/src/num/f128.rs:
```diff
@@ -229,6 +272,75 @@ impl f128 {
+    #[inline]
+    #[must_use]
+    #[cfg(not(bootstrap))]
+    #[stable(feature = "rust1", since = "1.0.0")]
```
Shouldn't this be `unstable(feature = "f128", issue = "116909")`? (And the same for the rest.)

tgross35 replied:
Thanks for the catch! I’m still working on this so just haven’t done a final look through yet

tgross35 replied:
This should be fixed now, feel free to take another look through

tgross35 committed:
Add doctests to existing `f16` and `f128` functions

The symbols that these tests rely on are not available on all platforms
and some ABIs are buggy, tests that rely on external functions are
configured to only run on x86 (`f128`) or aarch64 (`f16`).

tgross35 committed:
Add more `f16` and `f128` library functions and constants

This adds everything that was directly or transitively blocked on const
arithmetic for these types, which was recently merged.

Since const arithmetic is recent, most of these need to be gated by
`bootstrap`.

Anything that relies on intrinsics that are still missing is excluded.

tgross35 committed:
Add tests for `f16` and `f128`

This suite tests all library functions that are now available for the
types. Tests are only run on certain platforms where `f16` and `f128`
are known to work (have symbols available and don't crash LLVM).

tgross35 commented:
CI is still failing because of Clippy, so this is blocked on https://github.com/rust-lang/rust/pull/126636. However, I have tested locally on both x86_64 and aarch64 and everything passes (both stage0 and stage1), so I think this is good enough for a review.

There are lots of FIXMEs where we are missing some function or intrinsic for a specific test, and tests aren't even enabled on most platforms because LLVM support has been iffy, but this at least gets us off the ground with some end to end testing.

@rustbot ready
@rustbot label +F-f16_and_f128

tgross35 commented:
Whoever reviews please rollup=never, pretty high chance of hitting a bug on at least one platform

tgross35 commented on library/core/src/num/f128.rs:
```diff
@@ -220,21 +221,137 @@ impl f128 {
+    pub const fn is_infinite(self) -> bool {
+        // Getting clever with transmutation can result in incorrect answers on some FPUs
+        // FIXME: alter the Rust <-> Rust calling convention to prevent this problem.
+        // See https://github.com/rust-lang/rust/issues/72327
```
Thanks, I removed the comment

tgross35 commented:
Hm, getting some failures that look like ABI issues https://github.com/rust-lang/rust/actions/runs/9574955780/job/26399057121#step:25:3458. From the logs it looks like this should be `x86_64-unknown-linux-gnu`, but maybe it is using the host mingw symbols and running into https://gcc.gnu.org/bugzilla/show_bug.cgi?id=115054. I'm not sure how linking works there.

Any ideas @beetrees?

tgross35 commented:
Thanks for testing, I guess I can just drop the f16 changes from this PR if it is ready to go before https://github.com/rust-lang/rust/pull/125016. Or try to find some platforms where it works correctly to keep testing on (i586?)

tgross35 commented on library/core/src/num/f128.rs:
```diff
@@ -278,6 +392,220 @@ impl f128 {
+    /// let angle = std::f128::consts::PI;
+    ///
+    /// let abs_difference = (angle.to_degrees() - 180.0).abs();
+    /// # #[cfg(any(not(target_arch = "x86"), target_feature = "sse2"))]
```
Thanks, removed it

RalfJung commented on library/core/src/num/f128.rs:
```diff
@@ -278,6 +392,219 @@ impl f128 {
+    // #[unstable(feature = "float_next_up_down", issue = "91399")]
+    pub fn next_up(self) -> Self {
+        // We must use strictly integer arithmetic to prevent denormals from
+        // flushing to zero after an arithmetic operation on some platforms.
```
I assume that's pre-existing, but -- "must" is too strong here. These platforms are buggy, our floating point operations are expected to always behave as specified by IEEE. `next_up` is just being nice and tries to behave correctly on buggy platforms.

tgross35 replied:
Yeah, this just came from `f32` and `f64`. I'll try to reword it a bit

tgross35 replied:
Adjusted the wording if you get a chance to take a look

RalfJung commented on library/std/src/f128.rs:
```diff
@@ -32,4 +32,36 @@ impl f128 {
+        // We don't do this now because LLVM has lowering bugs for f128 math.
+
+        // SAFETY: this is the same as
+        Self::from_bits(self.to_bits() & !(1 << 127))
```
This could use SIGN_MASK, couldn't it?

tgross35 replied:
That's in the wrong library :( but I also don't need 45% of a safety comment....

tgross35 committed:
Add build.rs config for reliable `f16` and `f128`

There are some complexities about what platforms we can test f16 and
f128 on.  Put this in build.rs so we have an easy way to configure tests
with a single attribute, and keep it up to date.

RalfJung commented on library/core/src/num/f128.rs:
```diff
@@ -278,6 +400,214 @@ impl f128 {
+    #[unstable(feature = "f128", issue = "116909")]
+    // #[unstable(feature = "float_next_up_down", issue = "91399")]
+    pub fn next_up(self) -> Self {
+        // We use strictly integer arithmetic to avoid platform-specific behavior (e.g.
+        // denormals flushing to zero after arithmetic operations)
```
```suggestion
        // Some targets violate Rust's assumption of IEEE semantics, e.g. by flushing
        // denormals to zero. This is in general unsound and unsupported, but here
        // we do our best to still produce the correct result on such targets.
```

tgross35 replied:
Sure, added your wording to all four floats

tgross35 committed:
Reword docs for `f32` and `f64`

Better explain the reasoning for the `next_up`/`next_down` integer
implementation, as requested by Ralf.

tspiteri commented on library/core/src/num/f128.rs:
```diff
@@ -278,6 +400,216 @@ impl f128 {
+    #[must_use = "this returns the result of the operation, without modifying the original"]
+    pub fn to_degrees(self) -> Self {
+        // Use a literal for better precision.
+        const PIS_IN_180: f128 = 57.295779513082320876798154814105170332405472466564321549152_f128;
```
While this parses as the correct f128 number, the last few digits here are incorrect. If we use 60 significant digits like in the other f128 constants, it would end "491602_f128" instead of "49152_f128".

tgross35 replied:
These just do the truncation in binary precision (to 190 bits) before converting to decimal, rather than converting to decimal then truncating the string. But I updated them if that is consistent, mind checking f16 too?

tspiteri replied:
I had already checked f16 and they were fine.

tspiteri commented on library/core/src/num/f128.rs:
```diff
@@ -278,6 +400,216 @@ impl f128 {
+    pub fn to_radians(self) -> f128 {
+        // Use a literal for better precision.
+        const RADS_PER_DEG: f128 =
+            0.017453292519943295769236907684886127134428718885417254560967_f128;
```
Again, if we use 60 significant digits, this would end "609719_f128" instead of "60967_f128".

tgross35 commented on library/std/src/f128/tests.rs:
```diff
@@ -1,40 +1,541 @@
+
+// Ignore test on x87 floating point, these platforms do not guarantee NaN
+// payloads are preserved and flush denormals to zero, failing the tests.
+// #[cfg(not(target_arch = "x86"))]
```
Thanks for confirming, removed.

Mark-Simulacrum commented:
@bors r+

bors commented:
:pushpin: Commit a9bfab38e5c3b769789afe454f852feb750dd71d has been approved by `Mark-Simulacrum`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Mark-Simulacrum a9bfab38e5c3b769789afe454f852feb750dd71d -->
<!-- homu: {"type":"Approved","sha":"a9bfab38e5c3b769789afe454f852feb750dd71d","approver":"Mark-Simulacrum","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

compiler-errors commented:
if this ends up failing

@bors rollup=iffy

bors commented:
:hourglass: Testing commit a9bfab38e5c3b769789afe454f852feb750dd71d with merge 754aa90e92b0500b5f4ecb1161ba458ca0a8573e...
<!-- homu: {"type":"BuildStarted","head_sha":"a9bfab38e5c3b769789afe454f852feb750dd71d","merge_sha":"754aa90e92b0500b5f4ecb1161ba458ca0a8573e"} -->

rust-log-analyzer commented:
The job **`x86_64-msvc`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/9636189140/job/26573863015) [(plain)](https://github.com/rust-lang-ci/rust/commit/754aa90e92b0500b5f4ecb1161ba458ca0a8573e/checks/26573863015/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/9636189140/job/26574563149)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/9636189140/job/26574563149","builder_name":"checks-actions"} -->

compiler-errors commented:
@bors r- yeah it failed that rollup

tgross35 commented:
I updated the x86 doctests to also be linux-only, which is what I meant to do originally

tgross35 commented:
Had to disable f16 doctests on non-unix as well since it looks like extendhfsf/truncsfhf aren't always available. I was able to test locally on Windows MSVC and it passes there now.

@rustbot ready

Mark-Simulacrum commented:
@bors r+ rollup=never

bors commented:
:pushpin: Commit e181297c8cc814b2d3873b3a3b9d9d0182c660f1 has been approved by `Mark-Simulacrum`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Mark-Simulacrum e181297c8cc814b2d3873b3a3b9d9d0182c660f1 -->
<!-- homu: {"type":"Approved","sha":"e181297c8cc814b2d3873b3a3b9d9d0182c660f1","approver":"Mark-Simulacrum","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit e181297c8cc814b2d3873b3a3b9d9d0182c660f1 with merge 4bdf8d2d5877f20b54c1506a607ad8c4744cc387...
<!-- homu: {"type":"BuildStarted","head_sha":"e181297c8cc814b2d3873b3a3b9d9d0182c660f1","merge_sha":"4bdf8d2d5877f20b54c1506a607ad8c4744cc387"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/9679289592/job/26711964834)
Approved by: Mark-Simulacrum
Pushing 4bdf8d2d5877f20b54c1506a607ad8c4744cc387 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"Mark-Simulacrum","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/9679289592/job/26711964834"},"merge_sha":"4bdf8d2d5877f20b54c1506a607ad8c4744cc387"} -->

rust-timer commented:
Finished benchmarking commit (4bdf8d2d5877f20b54c1506a607ad8c4744cc387): [comparison URL](https://perf.rust-lang.org/compare.html?start=bae813a265014c22c37457ccf0d29fc6822b88da&end=4bdf8d2d5877f20b54c1506a607ad8c4744cc387&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=bae813a265014c22c37457ccf0d29fc6822b88da&end=4bdf8d2d5877f20b54c1506a607ad8c4744cc387&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=bae813a265014c22c37457ccf0d29fc6822b88da&end=4bdf8d2d5877f20b54c1506a607ad8c4744cc387&stat=max-rss)
<details>
<summary>Results (primary -2.8%, secondary 2.9%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 8.1%  | [8.1%, 8.1%]   | 1     |
| Improvements ✅ <br /> (primary)   | -2.8% | [-3.4%, -2.2%] | 2     |
| Improvements ✅ <br /> (secondary) | -2.2% | [-2.2%, -2.2%] | 1     |
| All ❌✅ (primary)                 | -2.8% | [-3.4%, -2.2%] | 2     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=bae813a265014c22c37457ccf0d29fc6822b88da&end=4bdf8d2d5877f20b54c1506a607ad8c4744cc387&stat=cycles:u)
<details>
<summary>Results (primary -1.5%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -1.5% | [-1.5%, -1.5%] | 1     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -1.5% | [-1.5%, -1.5%] | 1     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=bae813a265014c22c37457ccf0d29fc6822b88da&end=4bdf8d2d5877f20b54c1506a607ad8c4744cc387&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 695.46s -> 694.171s (-0.19%)
**Artifact size**: 326.69 MiB -> 324.67 MiB (-0.62%)
